diff/autofd/autofd.go

```diff
@@ -0,0 +1,373 @@
+			packages.NeedFiles |
+			packages.NeedCompiledGoFiles |
+			packages.NeedSyntax |
+			packages.NeedTypes |
+			packages.NeedTypesInfo,
+	}
+	pkgs, err := packages.Load(cfg, path)
+	if err != nil {
+		return nil, fmt.Errorf("could not load package of %q %s: %w", f.Path, f.Name, err)
+	}
+
+	var pkg *packages.Package
+	for _, p := range pkgs {
+		if p.PkgPath == path {
+			pkg = p
+			break
+		}
+	}
+
+	if pkg == nil || len(pkg.Errors) > 0 {
+		return nil, fmt.Errorf("could not find package %q", path)
+	}
+
+	var fct *types.Func
+	scope := pkg.Types.Scope()
+	switch {
+	case strings.Contains(name, "."):
+		idx := strings.Index(name, ".")
+		obj := scope.Lookup(name[:idx])
+		if obj == nil {
+			return nil, fmt.Errorf("could not find %s in package %q", name[:idx], path)
+		}
+		typ, ok := obj.Type().(*types.Named)
+		if !ok {
+			return nil, fmt.Errorf(
+				"object %s in package %q is not a named type (%T)",
+				name[:idx], path, obj,
+			)
+		}
+		for i := 0; i < typ.NumMethods(); i++ {
+			m := typ.Method(i)
+			if m.Name() == name[idx+1:] {
+				fct = m
+				break
+			}
+		}
+
+		if fct == nil {
+			return nil, fmt.Errorf("could not find %s in package %q", name, path)
+		}
+
+	default:
+		obj := scope.Lookup(name)
+		if obj == nil {
+			return nil, fmt.Errorf("could not find %s in package %q", name, path)
+		}
+		var ok bool
+		fct, ok = obj.(*types.Func)
+		if !ok {
+			return nil, fmt.Errorf("object %s in package %q is not a func (%T)", name, path, obj)
+		}
+	}
+
+	if !types.Identical(fct.Type(), f1x.Type()) {
+		return nil, fmt.Errorf("invalid function signature for %s", name)
+	}
+
+	der := f.Deriv
+	if der == "" {
+		der = "Deriv" + strings.Replace(f.Name, ".", "_", -1)
+	}
+
+	return &generator{w: w, pkg: pkg, fct: fct, rank: rank, der: der}, nil
+}
+
+func (g *generator) generate() error {
+	var fct *ast.FuncDecl
+	for _, f := range g.pkg.Syntax {
+		for i := range f.Decls {
+			decl, ok := f.Decls[i].(*ast.FuncDecl)
+			if !ok {
+				continue
+			}
+
+			if decl.Name.Name == g.fct.Name() {
+				fct = decl
+				break
+			}
+		}
+	}
+
+	var (
+		ret     *ast.ReturnStmt
+		returns int
+	)
+	ast.Inspect(fct.Body, func(n ast.Node) bool {
+		switch stmt := n.(type) {
+		case *ast.ReturnStmt:
+			returns++
+			ret = stmt
+		}
+		return true
+	})
+
+	switch returns {
+	case 0:
+		return fmt.Errorf("could not find a return statement")
+	case 1:
+		// ok
+	default:
+		return fmt.Errorf("can not handle functions with multiple return statements")
+	}
+
+	switch len(ret.Results) {
+	case 0:
+		return fmt.Errorf("naked returns not supported")
+	case 1:
+		// ok
+	default:
+		return fmt.Errorf("too many return values")
+	}
+
+	args := g.fct.Type().Underlying().(*types.Signature).Params()
+	g.printf("func %s(%s float64) float64 {\n",
+		g.der,
+		args.At(0).Name(),
+	)
+	g.printf("\tv := ")
+	g.expr(ret.Results[0])
+	switch g.rank {
+	case d1xKind:
+		g.printf("\n\treturn v.Emag\n")
+	case d2xKind:
+		g.printf("\n\treturn v.E1E2mag\n")
```

Comment:
I think this is a waste. The hyperdual returns both the first and second derivatives of the function at x; `g.printf("\n\treturn v.E1mag, v.E1E2mag\n")`. This is why I think that the parameterisation should be a `bool` indicating that the second derivative is invuded.

Reply:
my reasonning (but I admit I haven't benchmarked it) was that it would be CPU-wasteful to use `hyperdual` when only `dual` was needed.

if the overhead of `hyperdual` wrt `dual` is indeed minimal, then I guess that indeed we can indeed just generated `func Derivative(x float64) (d1, d2 float64) { ... }` and be done with it.
(and let users possibly wrap that function to only return one of the two values)

Reply:
I agree, but if you need the second derivative, you get the first for free. If you don't, you only use the dual. With the approach that's here, if you need both the first and second derivatives, you need to generate two functions, and you need to call both functions.